description: fix opening read-only files with FILE_DELETE_ON_CLOSE
author: Sebastian Lackner <sebastian@fds-team.de>
origin: https://github.com/wine-compholio/wine-staging/blob/master/patches/server-Delete_On_Close/0001-server-Fix-handling-of-opening-read-only-files-with-.patch

--- a/server/fd.c
+++ b/server/fd.c
@@ -1834,6 +1834,15 @@ struct fd *open_fd( struct fd *root, con
         list_add_head( &inode->open, &fd->inode_entry );
         closed_fd = NULL;
 
+        /* can't unlink files we don't have permission to access */
+        if ((options & FILE_DELETE_ON_CLOSE) && S_ISREG(st.st_mode) &&
+            !(flags & O_CREAT) && !(st.st_mode & (S_IWUSR | S_IWGRP | S_IWOTH)))
+        {
+            /* FIXME: instead of checking for O_CREAT it should check if the file was created */
+            set_error( STATUS_CANNOT_DELETE );
+            goto error;
+        }
+
         /* check directory options */
         if ((options & FILE_DIRECTORY_FILE) && !S_ISDIR(st.st_mode))
         {
